### PR TITLE
[Compute] `az vm list`: Add new paramter `--vmss` to support querying VM instances in a specific VMSS

### PR DESCRIPTION
**Pull request by PARADISSEEKR**
**Related command**

`az vm list`

**Description**

Add az vm list filter for VMSS
close https://github.com/Azure/azure-cli/issues/24609

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change
[Component Name 2] `az command b`: Add some customer-facing feature

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Compute

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 1]
```diff
@@ -458,6 +458,9 @@ def load_arguments(self, _):
         c.argument('port', help="The port or port range (ex: 80-100) to open inbound traffic to. Use '*' to allow traffic to all ports. Use comma separated values to specify more than one port or port range.")
         c.argument('priority', help='Rule priority, between 100 (highest priority) and 4096 (lowest priority). Must be unique for each rule in the collection.', type=int)
 
+    with self.argument_context('vm list') as c:
+        c.argument('vmss', min_api='2021-11-01', help='The system query option to filter VMs. Allowed value is vmss id.')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:462]: May I ask if it supports the name of VMSS?

Comment (fitzgeraldsteele) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py:462]: The API only accepts the full resource ID. I think there are two options:

1) Could CLI infer the vmss id from the parameters? 
`az vm list --vmss myvmss --resource-group demorg`

- That seems like the most customer friendly
- I think you would have to require the resource group if customer specifies `--vmss`. 
- You'd have to choose to make that call either at the subscription level or the resource group level (you can pass the filter to either)

2) Just make the parameter `--vmss-id`

- Simpler
- Doesn't seem as user friendly though

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/vm/_params.py:462]: OK, thanks for your good suggestion~

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 2]
```diff
@@ -8815,6 +8815,24 @@ def test_vm_create_count(self, resource_group):
...
+
+        self.cmd('vmss create -g {rg} -n {vmss_flex} --orchestration-mode Flexible --admin-username vmtest --platform-fault-domain-count 2 --image ubuntults --instance-count 2')
+        vmss_id = self.cmd('vmss show -g {rg} -n {vmss_flex}').get_output_in_json()['id']
+        self.kwargs.update({
+            'vmss_id': vmss_id
+        })
+        time.sleep(600)
+        self.cmd('vm list -g {rg} --vmss {vmss_id}', checks=[
+            self.check('length(@)', 2)
+        ])
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py:8833]: Could we add a test case for listing VM instances by VMSS name

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 3]
```diff
@@ -1353,10 +1353,23 @@ def list_skus(cmd, location=None, size=None, zone=None, show_all=None, resource_
...
+    from azure.cli.core.commands.client_factory import get_subscription_id
     ccf = _compute_client_factory(cmd.cli_ctx)
-    vm_list = ccf.virtual_machines.list(resource_group_name=resource_group_name) \
-        if resource_group_name else ccf.virtual_machines.list_all()
+    if vmss is not None:
+        if is_valid_resource_id(vmss):
+            filter = "'virtualMachineScaleSet/id' eq '{}'".format(vmss)
+        else:
+            vmss_id = resource_id(subscription=get_subscription_id(cmd.cli_ctx), resource_group=resource_group_name,
+                                  namespace='Microsoft.Compute', type='virtualMachineScaleSets', name=vmss)
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:1366]: When the `--vmss` passed in as a VMSS name, we should add verification to make the `--resource-group` parameter required

File: src/azure-cli/azure/cli/command_modules/vm/_params.py [hunk 4]
```diff
@@ -458,6 +458,9 @@ def load_arguments(self, _):
         c.argument('port', help="The port or port range (ex: 80-100) to open inbound traffic to. Use '*' to allow traffic to all ports. Use comma separated values to specify more than one port or port range.")
         c.argument('priority', help='Rule priority, between 100 (highest priority) and 4096 (lowest priority). Must be unique for each rule in the collection.', type=int)
 
+    with self.argument_context('vm list') as c:
+        c.argument('vmss', min_api='2021-11-01', help='The system query option to filter VMs. Allowed value is vmss id or vmss name.')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/_params.py:462]:
```suggestion
        c.argument('vmss', min_api='2021-11-01', help='List VM instances in a specific VMSS. Please specify the VMSS id or VMSS name.')
```

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 5]
```diff
@@ -1353,10 +1353,26 @@ def list_skus(cmd, location=None, size=None, zone=None, show_all=None, resource_
...
     ccf = _compute_client_factory(cmd.cli_ctx)
-    vm_list = ccf.virtual_machines.list(resource_group_name=resource_group_name) \
-        if resource_group_name else ccf.virtual_machines.list_all()
+    if vmss is not None:
+        if is_valid_resource_id(vmss):
+            filter = "'virtualMachineScaleSet/id' eq '{}'".format(vmss)
+        else:
+            if resource_group_name is None:
+                raise RequiredArgumentMissingError(
+                    'usage error: --resource-group-name required to list vm with vmss name')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:1367]:
```suggestion
                    'usage error: --resource-group required to list vm with vmss name')
```

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 6]
```diff
@@ -1353,10 +1353,26 @@ def list_skus(cmd, location=None, size=None, zone=None, show_all=None, resource_
...
+            vmss_id = resource_id(subscription=get_subscription_id(cmd.cli_ctx), resource_group=resource_group_name,
+                                  namespace='Microsoft.Compute', type='virtualMachineScaleSets', name=vmss)
+            filter = "'virtualMachineScaleSet/id' eq '{}'".format(vmss_id)
+        vm_list = ccf.virtual_machines.list(resource_group_name=resource_group_name, filter=filter) \
+            if resource_group_name else ccf.virtual_machines.list_all()
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:1372]: If users only pass in the `--vmss` without `--resource-group`, it should not query all the VMs

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 7]
```diff
@@ -8815,6 +8815,27 @@ def test_vm_create_count(self, resource_group):
...
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py:8833]: Lets add a test case for `vm list --vmss {vmss_id}`

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 8]
```diff
@@ -1353,10 +1353,27 @@ def list_skus(cmd, location=None, size=None, zone=None, show_all=None, resource_
...
-        if resource_group_name else ccf.virtual_machines.list_all()
+    if vmss is not None:
+        if is_valid_resource_id(vmss):
+            filter = "'virtualMachineScaleSet/id' eq '{}'".format(vmss)
+            if resource_group_name is None:
+                resource_group_name = parse_resource_id(vmss)['resource_group']
+        else:
+            if resource_group_name is None:
+                raise RequiredArgumentMissingError(
+                    'usage error: --resource-group required to list vm with vmss name')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:1369]:
```suggestion
                    'usage error: please specify the --resource-group when list VM instances with VMSS name')
```

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 9]
```diff
@@ -8815,6 +8815,33 @@ def test_vm_create_count(self, resource_group):
...
+        self.cmd('vm list --vmss {vmss_id}', checks=[
+            self.check('length(@)', 2)
+        ])
+        self.cmd('vm list -g {rg} --vmss {vmss_flex}', checks=[
+            self.check('length(@)', 2)
+        ])
+        message = 'usage error: --resource-group required to list vm with vmss name'
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py:8840]:
```suggestion
        message = 'usage error: please specify the --resource-group when listing VM instances with VMSS name'
```

File: src/azure-cli/azure/cli/command_modules/vm/custom.py [hunk 10]
```diff
@@ -1353,10 +1353,27 @@ def list_skus(cmd, location=None, size=None, zone=None, show_all=None, resource_
...
-        if resource_group_name else ccf.virtual_machines.list_all()
+    if vmss is not None:
+        if is_valid_resource_id(vmss):
+            filter = "'virtualMachineScaleSet/id' eq '{}'".format(vmss)
+            if resource_group_name is None:
+                resource_group_name = parse_resource_id(vmss)['resource_group']
+        else:
+            if resource_group_name is None:
+                raise RequiredArgumentMissingError(
+                    'usage error: please specify the --resource-group when list VM instances with VMSS name')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/custom.py:1369]:
```suggestion
                    'usage error: please specify the --resource-group when listing VM instances with VMSS name')
```

File: src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py [hunk 11]
```diff
@@ -8815,6 +8815,33 @@ def test_vm_create_count(self, resource_group):
...
+        self.cmd('vm list -g {rg} --vmss {vmss_id}', checks=[
+            self.check('length(@)', 2)
+        ])
+        self.cmd('vm list --vmss {vmss_id}', checks=[
+            self.check('length(@)', 2)
+        ])
+        self.cmd('vm list -g {rg} --vmss {vmss_flex}', checks=[
+            self.check('length(@)', 2)
+        ])
+        message = 'usage error: please specify the --resource-group when list VM instances with VMSS name'
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/vm/tests/latest/test_vm_commands.py:8840]:
```suggestion
        message = 'usage error: please specify the --resource-group when listing VM instances with VMSS name'
```

**Review by zhoxing-ms — APPROVED** [review]

**Comment by fitzgeraldsteele** [comment]
Awesome! Thank you, team!

**Pull request merged**